### ISSUE TITLE
RWD (Beta) Accessibility Anti-Patterns

### ISSUE THREAD

**Issue by bbsmooth** [freeCodeCamp/freeCodeCamp#44609]
## Affected page

[Learn More About CSS Pseudo Selectors By Building A Balance Sheet / Step 44](https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-more-about-css-pseudo-selectors-by-building-a-balance-sheet/step-44)
And by extension, steps 45 and 46.

## Discussion

There two accessibility anti-patterns in step 44:

A) Assigning a `tabindex` value greater than 0 to an element
B) Inserting a non-interactive element into the tab order (i.e. assigning a `tabindex` value greater than -1 to a non-interactive element)

Concerning A, the overwhelming majority of accessibility professionals will tell you to never use a `tabindex` higher than 0. Granted, because there are no real interactive elements in this particular course project then using `tabindex=1` on all three elements does not technically cause them to be placed out of the natural flow of keyboard focus (since they are the only elements in the flow). But this is a very contrived example and most likely in a real world application there would be other interactive elements on the page and thus setting `tabindex` to 1 would adversely affect keyboard accessibility. I fear that using `tabindex=1` here legitimizes its usage, even with the disclaimer. Most beginners are not going to have enough experience with accessibility to understand the ramifications of using a `tabindex` greater than 0. But really, it's an anti-pattern that should not be utilized in a teaching environment.

### References
[WebAIM: Keyboard Accessibility: Tabindex](https://webaim.org/techniques/keyboard/tabindex)
[Don’t Use Tabindex Greater than 0](https://adrianroselli.com/2014/11/dont-use-tabindex-greater-than-0.html)
[Elements should not have tabindex greater than zero](https://dequeuniversity.com/rules/axe/4.0/tabindex)
[Using tabindex](https://developers.google.com/web/fundamentals/accessibility/focus/using-tabindex)

Concerning B, while the intent of enhancing accessibility is noble, inserting these non-interactive elements into the tab order can actually reduce accessibility. Users, especially users of assistive technology have certain expectations about elements that can receive keyboard focus, primarily that they are interactive and provide some sort of functionality, which these elements do not. Also, a keyboard focusable element should have a proper name and interactive role defined in the accessibility tree, which these elements do not. The solution to enhancing accessibility in this particular example is not to add keyboard focus to the content but rather to make the content more semantically meaningful, most likely using HTML tables, which has the added benefit that screen readers will provide their own keyboard shortcuts for traversing the tables.

### References
[Use the tabindex attribute](https://www.a11yproject.com/posts/2021-01-28-how-to-use-the-tabindex-attribute/)
[MDN: Focusable elements should have interactive semantics](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Keyboard#focusable_elements_should_have_interactive_semantics)
[WebAIM: Keyboard Accessibility](https://webaim.org/techniques/keyboard/)
[no-noninteractive-tabindex](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-tabindex.md)

## Recommendation

Remove all references to tabindex in steps 44, 45, and 46. Step 44 merely adds the tabindex, so it could possibly be removed completely. Step 45 demonstrates the usage of the `:hover` pseudo class, which could still be done without the `tabindex` attribute. Step 46 demonstrates the usage of the `:focus` pseudo class. Since there are no real focusable elements on the page you would need to introduce some if you wanted to keep this step, which you could have the user add in step 44.

**Comment by ShaunSHamilton** [comment]
@bbsmooth As always, a huge thank you for taking the time to share your expertise.

What you have shared makes sense. Unfortunately, as Campers have likely no knowledge of JavaScript at this point, we cannot make use of some of the alternative recommendations.

I suppose we do not want to remove the usage/teaching of `:hover` or `:focus` or `tabindex`. I suggest we take this issue, and discuss from the beginning: _Should this project be completely re-worked?_

I am unsure why it does not use the HTML table elements. @nhcarrigan Can you remember?

**Comment by bbsmooth** [comment]
@ShaunSHamilton I would probably argue that we might not want to introduce `tabindex` in these new courses. Yes, it is a part of the HTML standard, but its legitimate use cases are quite limited and require JS, such as managing focus in SPAs or custom widgets, so it is going to be hard to use `tabindex` properly without adding a lot of complexity.

I can think of one example where JS is not needed and that's with making tables more accessible to keyboard users:

[Under-Engineered Responsive Tables](https://adrianroselli.com/2020/11/under-engineered-responsive-tables.html)

So if we did update this particular course to use tables then we could introduce `tabindex=0` to make the tables more accessible. But even then I think we would need to make it very clear that this usage is for a very specific accessibility issue.

**Comment by nhcarrigan** [comment]
Thanks for bringing this up @bbsmooth!

Yes, I agree that removing `tabindex` is a good approach. I'll be looking at modifying this project to use an HTML table, after discussion with Shaun, at some point this week. 🙂
